Shopify/cli
# Extensions refactor: Switch to new system for dev/deploy

### WHY are these changes introduced?
As part of the extensions revamp project, this PR switches the system for loading extensions: From the old plain object, we now create `ExtensionInstance` or `FunctionInstance`

For now, the change only applies for the `dev` and `deploy` flows, the `generate` flow still has a lot of hardcoded stuff. That will come in a different PR to keep this one smaller.

Contributes to https://github.com/Shopify/shopify-cli-planning/issues/439



### WHAT is this pull request doing?
- Main change: on `loader.ts`, create `ExtensionInstance` and `FunctionInstance` instances instead of inline plain objects.
- Update extensions/functions specs to add some missing properties
- Remove hardcoded extension switches where possible (still a few more pending!)
- Updated `UIExtension` and the other interfaces exposing new properties/methods from the instances, this means that a looot of tests needed to be updated. That's way there are so many changes too.

### How to test your changes?
- Tests that all extensions still work. These are the test I did manually:
  - Make sure the json output from the extension server is the same before and after this PR
  - Make sure all extensions work (even theme app extension)
  - Make sure all extensions can be deployed (including functions)



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
